### PR TITLE
Benchmarks: Fix Bug - Fix gemm kernel bug for nvidia v100.

### PR DESCRIPTION
**Pull request by guoshzhao**
**Description**
The order of changing the kernels and generating the commands is reversed.

### CODE REVIEW

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py [hunk 1]
```diff
@@ -106,10 +111,6 @@ def _preprocess(self):
                 self._commands.append(command)
 
         # TODO - To support more architecutres, currently only support compute capability = 7.0 or 8.0
-        capability = nv_helper.get_device_compute_capability()
-        if capability == 7.0:
-            self.__kernel_map['FP16_TC'] = 'cutlass_tensorop_h884gemm_256x128_32x2_*'
-
         if capability not in [7.0, 8.0]:
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:114]: Do we need to move all these check before command generation?

Comment (guoshzhao) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:114]: To better test the command generation code, I kept the unsupported architecture checking after the command generation code. So we can test it on any platform.

Comment (cp5555) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:114]: I think we should do it in the UT, not in the code. Correct?

**Review by cp5555 — APPROVED** [review]

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py [hunk 2]
```diff
@@ -24,16 +24,25 @@ def __init__(self, name, parameters=''):
 
         self._bin_name = 'cutlass_profiler'
 
+        # TODO - To support more architecutres, currently only support compute capability = 7.x or 8.x
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:27]: only support 7.0 and 8.0? 7.0 and 7.5 are different

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py [hunk 3]
```diff
@@ -24,16 +24,25 @@ def __init__(self, name, parameters=''):
...
-            'FP64': 'cutlass_simt_dgemm_128x128_8x2_*',
-            'FP32': 'cutlass_simt_sgemm_128x128_8x2_*',
-            'FP16': 'cutlass_simt_hgemm_256x128_8x2_*',
-            'FP64_TC': 'cutlass_tensorop_d884gemm_128x128_16x3_*',
-            'TF32_TC': 'cutlass_tensorop_tf32_s1688gemm_tf32_256x128_16x3_*',
-            'BF16_TC': 'cutlass_tensorop_bf16_s16816gemm_bf16_256x128_32x3_*',
-            'FP16_TC': 'cutlass_tensorop_h16816gemm_256x128_32x3_*',
-            'INT8_TC': 'cutlass_tensorop_s8_i16832gemm_s8_256x128_64x3_*',
-            'INT4_TC': 'cutlass_tensorop_s4_i16864gemm_s4_256x128_128x3_*',
+            7: {
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:29]: 7.5?

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py [hunk 4]
```diff
@@ -85,40 +94,48 @@ def _preprocess(self):
...
             self._result.set_return_code(ReturnCode.MICROBENCHMARK_UNSUPPORTED_ARCHITECTURE)
             logger.error(
-                'Unsupported architecture - benchmark: {}, compute capability: {}, expected: 7.0 or 8.0'.format(
+                'Unsupported architecture - benchmark: {}, compute capability: {}, expected: 7.x or 8.x'.format(
                     self._name, capability
                 )
             )
             return False
 
+        arch = int(capability)
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:108]: int is not enough, need to check both major and minor

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py (hunk 2 above)
Comment (abuccts) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:27]: e.g., Volta refers to sm7.0, Turing refers to sm7.5, they have very different tensor cores.

Comment (guoshzhao) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:27]: Do you know how to get the exact kernel name for different arch? Can we say that same arch have same kernel name?
For example, both 8.0 and 8.6 are ampere, do they have same gemm kernels?

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py (hunk 3 above)
Comment (guoshzhao) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:29]: 7.0 for V100?

Comment (abuccts) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py]: sorry, V100 is 7.0

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py (hunk 2 above)
Comment (abuccts) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py]: nope, not sure about that, but 7.5 introduces int tensor core which 7.0 does not have.
for sm8.0 and sm8.6, maybe need to compile cutlass profiler on both arch then check generated kernels?

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py [hunk 5]
```diff
@@ -72,8 +81,8 @@ def add_parser_arguments(self):
             '--precision',
             type=str,
             nargs='+',
-            default=list(self.__kernel_map.keys()),
-            help='Precision for benchmarking. E.g. {}.'.format(' '.join(list(self.__kernel_map.keys()))),
+            default=list(),
+            help='Precision for benchmarking. E.g. {}.'.format(' '.join(list(self.__kernel_map[8].keys()))),
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:85]: change to 8.0 accordingly?

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py [hunk 6]
```diff
@@ -85,32 +94,9 @@ def _preprocess(self):
...
-                self._commands.append(command)
-
-        # TODO - To support more architecutres, currently only support compute capability = 7.0 or 8.0
+        # Reset kernels according to compute capability.
         capability = nv_helper.get_device_compute_capability()
-        if capability == 7.0:
-            self.__kernel_map['FP16_TC'] = 'cutlass_tensorop_h884gemm_256x128_32x2_*'
-
-        if capability not in [7.0, 8.0]:
+        if capability is None or capability not in self.__kernel_map:
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:99]: no need for `capability is None`?

**Review by abuccts — APPROVED** [review]

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py (hunk 5 above)
Comment (guoshzhao) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py:85]: good catch.

File: superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py (hunk 6 above)
Comment (guoshzhao) [reply superbench/benchmarks/micro_benchmarks/gemm_flops_performance.py]: done.

**Pull request merged**